crmprtd/insert.py

```diff
@@ -142,85 +178,133 @@ def bisect_insert_strategy(sesh, obs):
-
+    elif len(observations) == 1:
+        return insert_single_obs(sesh, observations[0])
     # The happy case: add everything at once
     else:
         try:
             with sesh.begin_nested():
-                log.debug("New SAVEPOINT", extra={"num_obs": len(obs)})
-                sesh.add_all(obs)
+                log.debug("New SAVEPOINT", extra={"num_obs": len(observations)})
+                sesh.add_all(observations)
         except IntegrityError:
             log.debug("Failed, splitting observations.")
             sesh.rollback()
 
-            a, b = split(obs)
+            a, b = split(observations)
             dbm_a = bisect_insert_strategy(sesh, a)
             dbm_b = bisect_insert_strategy(sesh, b)
             return dbm_a + dbm_b
         else:
             log.info(
-                f"Successfully inserted observations: {len(obs)}",
-                extra={"num_obs": len(obs)},
+                f"Successfully inserted observations: {len(observations)}",
+                extra={"num_obs": len(observations)},
             )
-            db_metrics = DBMetrics(len(obs), 0, 0)
+            db_metrics = DBMetrics(len(observations), 0, 0)
         sesh.commit()
         return db_metrics
 
 
-def insert(sesh, observations, sample_size):
+def bulk_insert_strategy(sesh, observations):
+    """
+    This method performs a bulk insert of observations using the PostgreSQL dialect
+    INSERT ... ON CONFLICT DO IGNORE clause that allows bulk inserts with duplicates
+    to proceed without raising an exception (dups are ignored; new rows are inserted).
+
+    NOTE: This only works for PostgreSQL databases. Other dialects may support a
+    similar operation, but this code does not handle those cases.
+
+    :param sesh: SQLAlchemy database session
+    :param observations: List of observations to insert
+    :return: DMMetrics describing result of insertion
+    """
+    num_to_insert = len(observations)
+
+    # If you try to insert no rows with pg_insert, there's an unexpected insertion.
+    # Avoid that.
+    if num_to_insert == 0:
+        return DBMetrics(0, 0, 0)
+
+    try:
+        with sesh.begin_nested():
+            result = sesh.execute(
+                pg_insert(Obs)
+                .values(observations)
+                .on_conflict_do_nothing()
+                .returning(Obs.id)
+            ).fetchall()
+    except DBAPIError as e:
+        # Something really unanticipated happened. Duplicate rows do not trigger an
+        # exception.
+        log.warning(f"Unexpected error during bulk insertion: {e}")
+        return DBMetrics(0, 0, num_to_insert)
+    else:
+        num_inserted = len(result)
+        log.info(
+            f"Successfully inserted observations: {num_inserted}",
+            extra={"num_obs": num_inserted},
+        )
+    sesh.commit()
+    return DBMetrics(num_inserted, num_to_insert - num_inserted, 0)
+
+
+def insert(
+    sesh,
+    observations,
+    strategy=InsertStrategy.BULK,
```

Comment:
Part of what I liked about the old insert code was that it was adaptive and didn't require user control (only user *tuning*) over which insert strategy was used. If it detected a large percentage of duplicates, single insert was faster, so it would use that. Otherwise it would use the bulk/bisect strategy.

So that raises the question to me as to whether there are any problematic scenarios when your new bulk insert strategy *wouldn't* be faster. I suspect not, in which case the adaptiveness is now irrelevant, so making it user specified (with a sane default) is fine. But do you see any potential pitfalls?

That said, the bisect insert strategy and the single insert strategy are separate and you have conflated them into a hybrid of select-both-and-get-one-or-the-other. If we're going to commit to requiring user selection, I'd add single-insert as an option and just pull out the adaptive sampling code.

Reply:
I thought about this, and decided not to mess with any of the existing code, apart from making it optional. My mistake here was naming the alternative "BISECTION". It should have remained "OTHER" as it was earlier -- but that seemed too vague.

I haven't been able to foresee any circumstances under which the new bulk insert strategy would be slower. It requires precisely one round trip per set of rows to insert. Single insert and bisection all require more than one, except in the case that a large block of bisectable observations inserts successfully the first time. There's evidently a significant overhead to a round trip (and I don't know why).

Given that adaptive inserting avoids the worts case in which bisection bisects continually down to single inserts (which would require _log n_ insert attempts before succeeding or failing on a single insert), I don't think this it's a good idea to separate them ... unless for purposes of controlling comparisons, which might not be a bad idea. How about this:
- BULK (or some better name): INSERT ... ON CONFLICT DO NOTHING
- SINGLE: single insert
- BISECTION: ...
- ADAPTIVE: the old adaptive way, currently selected by BISECTION (bad name).

Reply:
Going ahead with this, which covers all bases. Can be modified later if necessary.